Keysequences

hexa00 opened the pull request:
Support keysequences
    
    This patch enables support for key sequences so that for example
    keybindings like this one are supported:
    
    "Ctrl-u t", meaning press Ctrl-u and then press t.
    
    Here's a few limitations at the moment however:
    
      - These keybindings can't be shown as part of an electron MenuItem see:
      https://github.com/electron/electron/issues/11740#issuecomment-360980460
    
      - Monaco only supports a sequence of 2 keybinding so "Ctrl-u t" will
        work but not "Ctrl-u t a"
    
        This means that for now in the QuickMenu if the sequence is more then
        2 keychords long we don't show it.
    
    Also note that at the moment the delay to enter a sequence is hardcoded to 1 sec
    between keystrokes.
    
    Fixes: #472

Note I will create issues about these limitations and see what can be done.

akosyakov commented on packages/core/src/common/keys.ts:
```diff
@@ -9,30 +9,98 @@ import { isOSX } from './os';
+        return true;
+    }
+
+    export enum compareResult {
```
It should be in the camel case

hexa00 replied:
OK Will change to CompareResult

hexa00 commented:
@akosyakov  yes see: registerKeybinding takes care of warning about such shadowing and doesn't allow the keybinding.

akosyakov commented on packages/core/src/common/keybinding.ts:
```diff
@@ -311,37 +338,66 @@ export class KeybindingRegistry {
-                    const command = this.commandRegistry.getCommand(binding.command);
-                    if (command) {
-                        const commandHandler = this.commandRegistry.getActiveHandler(command.id);
+        if (bindings.full.length > 0) {
```
I think it will be better to extract a function that returns true if a command was executed and false if not, in the second case we will try to match partial bindings

hexa00 replied:
OK it does make for a long function here.

akosyakov replied:
I was more concern about mixing code of execution command and handling partial bindings, such function won't need to care about them.
```js
if (fn(bindings)) {
    this.keySequence = [];
} else if (bindings.partial.length > 0) {
    ...
}
```

hexa00 replied:
yes it helps with that too, I'll refactor that a bit

akosyakov commented on packages/core/src/common/keybinding.ts:
```diff
@@ -311,37 +338,66 @@ export class KeybindingRegistry {
+                    /* Try to match a partial binding.  */
+                    if (bindings.partial.length > 0) {
+                        this.keySequence.push(keyCode);
+                        this.timeoutTimer = setTimeout(() => {
```
Do we need a timeout? I see that VS code notifies a user about the current sequence in the status bar and that it is waiting for next character. If nothing is matched eventually it tells that is no such command for such key combination.

hexa00 replied:
Well the timeout is the would  to do like vscode and  `eventually tell that there is no such command` the eventually is the timeout.

I think it's good to modify the status bar but I would make that in another PR as well as having a preference for the timeout as the preference-api is merged in core.

akosyakov replied:
ok, should we have another issue for it?

hexa00 replied:
Yes I'll open it with the others about the limitations once this is merged.

akosyakov commented on packages/core/src/common/keybinding.ts:
```diff
@@ -150,16 +154,25 @@ export class KeybindingRegistry {
+            if (existingBindings.partial.length > 0) {
+                const shadows = existingBindings.partial.filter(b => b.context === binding.context);
+                if (shadows.length > 0) {
+                    this.logger.warn('Shadowing keybinding is ignored; ',
```
I thought shadowing keybinding means that a keybinding from one scope shadows a keybinding from another. I think here 2 keybindings are colliding again or overlapping. Do I miss something?

hexa00 replied:
It depends on your definition of shadowing... I've used the term to describe that if you had `ctrl-x ctrl-t` and try to register `ctrl-x` then it would shadow `ctrl-x ctrl-t` since you could not get to that keybinding.  I could use the word colliding too but that to me meant more that it was using the same exact keybinding...

hexa00 replied:
@akosyakov you will get a warning in that case too since if it will check do I have a keybinding for `ctrl-x ctrl-t` and it will find a partial keybinding for it in `ctrl-x` and refuse to register it. In this case `ctrl-x` would shadow `ctrl-x ctrl-t` ... but I think I have a problem if you have `ctrl-x ctrl-t` and try to regiter `ctrl-x ctrl-a` you'll get in the same conditions :( I have to fix that...

akosyakov commented on packages/core/src/common/keybinding.ts:
```diff
@@ -192,29 +205,43 @@ export class KeybindingRegistry {
      */
-    getKeybindingsForKeyCode(keyCode: KeyCode): Keybinding[] {
-        const result: Keybinding[] = [];
+    getKeybindingsForKeySequence(keySequence: KeySequence): { full: Keybinding[], partial: Keybinding[] } {
```
for clients, it would be good to have a named type

hexa00 replied:
Yes I hesitated about that.. I'll add this then.

hexa00 commented:
@akosyakov About Atl on OSX i wish I could test that there's been reports that ctrl doesn't work either see : #1060 

From the code I can't see what would be wrong there... 

Could anyone with mac dig deeper into this please ? from master to start with ?

hexa00 commented:
@akosyakov review comments have been addressed in this new version:

 - Named type for KeybindingsResult
 - Refactored run function with tryKeybindingExecution
 - Fixed & added test for `ctrl-x ctrl-a` that would wrongly conflict with `ctrl-x ctrl-b`

akosyakov commented on packages/core/src/common/keybinding.ts:
```diff
@@ -337,12 +367,48 @@ export class KeybindingRegistry {
-                break;
+                return false;
+            } else {
+                return false;
```
Should not we try next binding here?

hexa00 replied:
Oops indeed fixed. thx!

kittaakos commented:
> I've noticed that no keybindings work if they contain alt on mac. I am not sure that it is related to this PR.

@akosyakov, do you have a broken example that contains the `alt` modifier? I am fixing keybinding issues now, and I was unable to reproduce the defect you have described. Maybe I have already fixed it... Thanks!

akosyakov commented:
no, if it works now for you it is fine, I will test your PR with all modifiers :)
